tidy3d/components/geometry.py

```diff
@@ -3105,8 +3106,108 @@ def _surface_area(self, bounds: Bound) -> float:
         return area
 
 
+class CustomSurfaceGeometry(Geometry, ABC):
+    """Custom surface geometry given by a triangle mesh, as in the STL file format.
+
+    Example
+    -------
+    >>> from stl import mesh
+    >>> # import STL
+    >>> stl_mesh = mesh.Mesh.from_file("import.stl")
+    >>> geometry = CustomSurfaceGeometry.from_stl(stl_mesh.vectors)
+    >>> # export STL
+    >>> vectors = geometry.to_stl()
+    >>> stl_mesh = mesh.Mesh(np.zeros(len(vectors), dtype=mesh.Mesh.dtype))
+    >>> stl_mesh.vectors = vectors
+    >>> stl_mesh.save("export.stl")
+    """
+
+    triangle_data: TriangleDataArray = pydantic.Field(
+        ...,
+        title="Surface triangle vertices",
+        description="Data array consisting of surface triangle vertices.",
+    )
+
+    @classmethod
+    def from_stl(cls, triangle_vertices: ArrayLike[3]) -> CustomSurfaceGeometry:
+        """Load a :class:`.CustomSurfaceGeometry` from the data obtained from an STL file.
+
+        Parameters
+        ----------
+        triangle_vertices : ArrayLike[3]
+            An N x 3 x 3 array of the vertices of the triangles of the surface geometry.
+            The first index labels the triangle; the second index labels the coordinate
+            (x, y, or z); and the third index labels the vertex within a triangle
+            (0, 1, or 2).
+
+        Returns
+        -------
+        :class:`.CustomSurfaceGeometry`
+            The custom surface geometry given by the triangle vertices provided.
+        """
+        num_triangles = len(triangle_vertices)
+        coords = dict(
+            triangle_index=np.arange(num_triangles),
+            coordinate=np.arange(3),
```

Comment:
maybe double check this for consistency with STL format because in meshing schemes that I'm familiar with, the coordinate (actual position of the vertex) comes last and the vertex index (0,1,2) comes 2nd.

Reply:
you're right, thanks